File: domain/src/main/java/no/nav/melosys/domain/brev/DoksysBrevbestilling.java
package no.nav.melosys.domain.brev;

import java.util.Arrays;
import java.util.Collection;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.kodeverk.Aktoersroller;
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter;

@JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION, defaultImpl = DoksysBrevbestilling.class)
public class DoksysBrevbestilling extends Brevbestilling {
    private Aktoersroller mottakerRolle;
    private Collection<Mottaker> mottakere; //NOTE Flytt opp til Brevbestilling
    private String begrunnelseKode;
    private String fritekst;
    private String ytterligereInformasjon;

    public DoksysBrevbestilling() {
        super();
        //Tom constructor på grunn av deserialsering i prosessinstans
    }

    protected DoksysBrevbestilling(Produserbaredokumenter produserbartdokument,
                                   String avsenderID,
                                   Aktoersroller mottakerRolle,
                                   Collection<Mottaker> mottakere,
                                   Behandling behandling,
                                   String begrunnelseKode,
                                   String fritekst,
                                   String ytterligereInformasjon) {
        super(produserbartdokument, behandling, avsenderID);
        this.mottakerRolle = mottakerRolle;
        this.mottakere = mottakere;
        this.begrunnelseKode = begrunnelseKode;
        this.fritekst = fritekst;
        this.ytterligereInformasjon = ytterligereInformasjon;
    }

    public Aktoersroller getMottakerRolle() {
        return mottakerRolle;
    }

    public Collection<Mottaker> getMottakere() {
        return mottakere;
    }

    public String getBegrunnelseKode() {
        return begrunnelseKode;
    }

    public String getFritekst() {
        return fritekst;
    }

    public String getYtterligereInformasjon() {
        return ytterligereInformasjon;
    }

    public static class Builder {
        private Produserbaredokumenter produserbartdokument;
        private String avsenderID;
        private Aktoersroller mottakerRolle;
        private Collection<Mottaker> mottakere;
        private Behandling behandling;
        private String begrunnelseKode;
        private String fritekst;
        private String ytterligereInformasjon;

        public Builder medProduserbartDokument(Produserbaredokumenter produserbartdokument) {
            this.produserbartdokument = produserbartdokument;
            return this;
        }

        public Builder medAvsenderID(String avsenderID) {
            this.avsenderID = avsenderID;
            return this;
        }

        public Builder medMottakerRolle(Aktoersroller mottakerRolle) {
            this.mottakerRolle = mottakerRolle;
            return this;
        }


        public Builder medMottakere(Mottaker... mottakere) {
            this.mottakere = Arrays.asList(mottakere);
            return this;
        }

        public Builder medMottakere(Collection<Mottaker> mottakere) {
            this.mottakere = mottakere;
            return this;
        }

        public Builder medBehandling(Behandling behandling) {
            this.behandling = behandling;
            return this;
        }

        public Builder medBegrunnelseKode(String begrunnelseKode) {
            this.begrunnelseKode = begrunnelseKode;
            return this;
        }

        public Builder medFritekst(String fritekst) {
            this.fritekst = fritekst;
            return this;
        }

        public Builder medYtterligereInformasjon(String ytterligereInformasjon) {
            this.ytterligereInformasjon = ytterligereInformasjon;
            return this;
        }

        public DoksysBrevbestilling build() {
            return new DoksysBrevbestilling(
                produserbartdokument,
                avsenderID,
                mottakerRolle,
                mottakere,
                behandling,
                begrunnelseKode,
                fritekst,
                ytterligereInformasjon
            );
        }
    }

}

File: domain/src/main/kotlin/no/nav/melosys/domain/brev/DoksysBrevbestilling.kt
package no.nav.melosys.domain.brev

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.arkiv.Distribusjonstype
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter

class DoksysBrevbestilling : Brevbestilling {
    val mottakerRolle: Mottakerroller?
    var mottakere: Collection<Mottaker>?
    val begrunnelseKode: String?
    val fritekst: String?
    val distribusjonstype: Distribusjonstype?
    val ytterligereInformasjon: String?

    @JsonCreator
    private constructor(
        @JsonProperty("produserbartdokument") produserbartdokument: Produserbaredokumenter?,
        @JsonProperty("avsenderID") avsenderID: String?,
        @JsonProperty("mottakerRolle") mottakerRolle: Mottakerroller?,
        @JsonProperty("mottakere") mottakere: Collection<Mottaker>?,
        @JsonProperty("behandling") behandling: Behandling?,
        @JsonProperty("begrunnelseKode") begrunnelseKode: String?,
        @JsonProperty("fritekst") fritekst: String?,
        @JsonProperty("ytterligereInformasjon") ytterligereInformasjon: String?,
        @JsonProperty("distribusjonstype") distribusjonstype: Distribusjonstype?
    ) : super(produserbartdokument, behandling, avsenderID) {
        this.mottakerRolle = mottakerRolle
        this.mottakere = mottakere
        this.begrunnelseKode = begrunnelseKode
        this.fritekst = fritekst
        this.ytterligereInformasjon = ytterligereInformasjon
        this.distribusjonstype = distribusjonstype
    }

    fun settMottaker(mottaker: Mottaker?) {
        mottakere = mottaker?.let { listOf(it) }
    }

    class Builder {
        private var produserbartdokument: Produserbaredokumenter? = null
        private var avsenderID: String? = null
        private var mottakerRolle: Mottakerroller? = null
        private var mottakere: Collection<Mottaker>? = null
        private var behandling: Behandling? = null
        private var begrunnelseKode: String? = null
        private var fritekst: String? = null
        private var distribusjonstype: Distribusjonstype? = null
        private var ytterligereInformasjon: String? = null

        fun medProduserbartDokument(produserbartdokument: Produserbaredokumenter?) = apply { this.produserbartdokument = produserbartdokument }

        fun medAvsenderID(avsenderID: String?) = apply { this.avsenderID = avsenderID }

        fun medMottakerRolle(mottakerRolle: Mottakerroller?) = apply { this.mottakerRolle = mottakerRolle }

        fun medMottakere(vararg mottakere: Mottaker) = apply { this.mottakere = mottakere.toList() }

        fun medMottakere(mottakere: Collection<Mottaker>?) = apply { this.mottakere = mottakere }

        fun medBehandling(behandling: Behandling?) = apply { this.behandling = behandling }

        fun medBegrunnelseKode(begrunnelseKode: String?) = apply { this.begrunnelseKode = begrunnelseKode }

        fun medFritekst(fritekst: String?) = apply { this.fritekst = fritekst }

        fun medDistribusjonsType(distribusjonstype: Distribusjonstype?) = apply { this.distribusjonstype = distribusjonstype }

        fun medYtterligereInformasjon(ytterligereInformasjon: String?) = apply { this.ytterligereInformasjon = ytterligereInformasjon }

        fun build() = DoksysBrevbestilling(
            produserbartdokument = produserbartdokument,
            avsenderID = avsenderID,
            mottakerRolle = mottakerRolle,
            mottakere = mottakere,
            behandling = behandling,
            begrunnelseKode = begrunnelseKode,
            fritekst = fritekst,
            ytterligereInformasjon = ytterligereInformasjon,
            distribusjonstype = distribusjonstype
        )
    }
}
